Additional loo utilities

LeeviLindgren opened the pull request:
Hi!

This PR adds functions to compute additional LOO utilities: mean absolute error, mean squared error, and root mean squared error for regression problems and accuracy and balanced accuracy for binary classification. I have already discussed this with @avehtari and opened a PR for more discussion and feedback.

avehtari commented:
I think this PR could be merged adding just the loo_predictive_error function, but it would be useful to think the bigger picture and how we could support loo_compare with different criteria

jgabry commented:
Looking at this now. Sorry for the delay!

jgabry commented:
@LeeviLindgren Thanks for the PR! Sorry for the delay in reviewing it. I'll make a few review comments now, but can you also edits from maintainers? You can do that by following the instructions here:  https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+#' # matrix and draws from the posterior predictive distribution
+#' library("rstanarm")
+#'
+#' # data from help("lm")
+#' d <- data.frame(
+#'   weight = c(ctl, trt),
```
This currently results in an error because `ctl` and `trt` are not defined.

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+#' )
+#' fit <- stan_glm(weight ~ group, data = d, refresh = 0)
+#' ll <- log_lik(fit)
+#' r_eff <- relative_eff(exp(-log_ratios), chain_id = rep(1:4, each = 1000))
```
`log_ratios` isn't defined anywhere so this will error

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+
+    predictive_error_fun <- .loo_predictive_error_fun(pred_error)
+
+    return(predictive_error_fun(y, pred_loo))
```
```suggestion
    predictive_error_fun(y, pred_loo)
```

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+  stopifnot(length(y) == length(yhat))
+  n <- length(y)
+  e <- abs(y - yhat)
+  return(list(estimate = mean(e), se = sd(e) / sqrt(n)))
```
```suggestion
  list(estimate = mean(e), se = sd(e) / sqrt(n))
```

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+  stopifnot(length(y) == length(yhat))
+  n <- length(y)
+  e <- (y - yhat)^2
+  return(list(estimate = mean(e), se = sd(e) / sqrt(n)))
```
```suggestion
  list(estimate = mean(e), se = sd(e) / sqrt(n))
```

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+  yhat <- as.integer(yhat > 0.5)
+  acc <- as.integer(yhat == y)
+  est <- mean(acc)
+  return(list(estimate = est, se = sqrt(est * (1-est) / n) ))
```
```suggestion
  list(estimate = est, se = sqrt(est * (1-est) / n) )
```

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+  bls_acc <- (tp + tn) / 2
+  # This approximation has quite large bias for small samples
+  bls_acc_var <- (tp * (1 - tp) + tn * (1 - tn)) / 4
+  return(list(estimate = bls_acc, se = sqrt(bls_acc_var / n)))
```
```suggestion
  list(estimate = bls_acc, se = sqrt(bls_acc_var / n))
```

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+loo_predictive_error.default <-
+  function(x,
+           y,
+           ll,
```
I wonder if we should call the argument `log_lik` instead of `ll` since for `psis()` we have an argument `log_ratios` and not `lr`. Not a big deal either way. @avehtari What do you think?

avehtari replied:
longer name is more descriptive, and thus better

jgabry replied:
Ok let's go with `log_lik`. @LeeviLindgren Can you update the argument name?

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,200 @@
+  function(x,
+           y,
+           ll,
+           pred_error = c("mae", "rmse", "mse", "acc", "balanced_acc"),
```
What do you (and @avehtari) think about the name `metric` instead of `pred_error`?

avehtari replied:
Good idea, I like `metric`, especially as "acc" is not an error

LeeviLindgren replied:
I think it would also then make sense to call the function `loo_predictive_metric`, any opinions on this?

jgabry replied:
Yeah I suppose if we're including metrics that aren't errors then `loo_predictive_error` isn't the best name.

jgabry reviewed (commented):
This looks great, thanks! I made a bunch of comments/suggestions but they're all really minor.

LeeviLindgren commented:
> @LeeviLindgren Thanks for the PR! Sorry for the delay in reviewing it. I'll make a few review comments now, but can you also edits from maintainers? You can do that by following the instructions here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Thanks for the comments, I will take a look! Edits from maintainers should be allowed.

LeeviLindgren commented:
@avehtari also suggested that it would be nice to support model comparison using these additional metrics. I suggest we try to merge this PR which implements the metrics, and I can start working on the functionality to support comparison. The initial idea I had was to write a method for `loo_compare` for `loo_predictive_error`-class (we can also come up with another name). What do you think @jgabry?

jgabry commented:
Thanks for the updates. Waiting to hear from @avehtari on the couple of questions in my review comments (about the argument names) but after that we can merge this (assuming the GitHub Actions checks pass). 

> @avehtari also suggested that it would be nice to support model comparison using these additional metrics. I suggest we try to merge this PR which implements the metrics, and I can start working on the functionality to support comparison. The initial idea I had was to write a method for `loo_compare` for `loo_predictive_error`-class (we can also come up with another name). Users could then pass output from the calls to `loo_predictive_error` to `loo_compare`. What do you think @jgabry?

Yeah I agree it would be nice to be able to do model comparison with these. My initial thought is that your suggestion of using a class and a `loo_compare` method seems like a good approach.

avehtari commented:
I agree with @jgabry's suggestions

jgabry commented on R/loo_predictive_error.R:
```diff
@@ -0,0 +1,202 @@
+#' )
+#' fit <- stan_glm(weight ~ group, data = d, refresh = 0)
+#' ll <- log_lik(fit)
+#' r_eff <- relative_eff(exp(-ll), chain_id = rep(1:4, each = 1000))
```
I think the doc needs to be regenerated one more time so that the .Rd file includes this change to `-ll` instead of `log_ratios`. That's why R cmd check is failing but the unit tests are passing.

LeeviLindgren commented:
- I changed the function to `loo_predictive_metric`
- Updated the wording in docs to match the updates
- `ll` argument is now called `log_lik`
- `pred_error` argument is now called `metric`

jgabry commented:
Looks good, thanks! Merging now.
